### PR TITLE
Perform nav bar computation in two passes

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/2065682

This follows the same change made in taggers back in https://github.com/dotnet/roslyn/pull/72878.

The idea is fairly simple.  When computing nav bar items, we always first try a simple and quick pass using frozen-partial semantics.  This allows the nav bars to be reasonably up to date with whatever the user jsut types, as well as with the last contents of SG documents and skeleton references.

As long as new changes are coming in, we keep performing those cheap computation passes.

When a cheap computation pass finishes, we then enqueue a request to do one final expensive pass that will actually update hte navbars with non-frozen data.  Thus, ensuring that we always finally get into a final, correct, state.  That expensive pass will be pre-empted by more cheap requests that come in.  But we'll always eventually run that expensive/correct pass as the final thing we do.

### CODE REVIEW

File: src/EditorFeatures/Core/NavigationBar/NavigationBarController.cs [hunk 1]
```diff
@@ -196,7 +206,13 @@ private void StartModelUpdateAndSelectedItemUpdateTasks()
...
             return;
 
-        _computeModelQueue.AddWork(default(VoidResult));
+        // Cancel any expensive, in-flight, nav--bar work as there's now a request to perform lightweight tagging. Note:
+        // intentionally ignoring the return value here.  We're enqueuing normal work here, so it has no associated
+        // token with it.
+        _ = _nonFrozenComputationCancellationSeries.CreateNext();
+        _computeModelQueue.AddWork(
+            new NavigationBarQueueItem(FrozenPartialSemantics: true, NonFrozenComputationToken: null),
+            cancelExistingWork: true);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/NavigationBar/NavigationBarController.cs:215]: mirrors what we do in tagging.

File: src/EditorFeatures/Core/NavigationBar/NavigationBarController_ModelComputation.cs [hunk 2]
```diff
@@ -24,7 +22,52 @@ internal partial class NavigationBarController
...
-    private async ValueTask<NavigationBarModel?> ComputeModelAndSelectItemAsync(ImmutableSegmentedList<VoidResult> _, CancellationToken cancellationToken)
+    private async ValueTask<NavigationBarModel?> ComputeModelAndSelectItemAsync(
+        ImmutableSegmentedList<NavigationBarQueueItem> queueItems, CancellationToken cancellationToken)
+    {
+        // If any of the requests are for frozen partial, then we do compute with frozen partial semantics.  We
+        // always want these "fast but inaccurate" passes to happen first.  That pass will then enqueue the work
+        // to do the slow-but-accurate pass.
+        var frozenPartialSemantics = queueItems.Any(t => t.FrozenPartialSemantics);
+
+        if (!frozenPartialSemantics)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/NavigationBar/NavigationBarController_ModelComputation.cs:33]: mirrors the same logic in tagging.

File: src/EditorFeatures/Core/NavigationBar/NavigationBarController_ModelComputation.cs [hunk 3]
```diff
@@ -24,7 +22,52 @@ internal partial class NavigationBarController
...
+            }
+            catch (OperationCanceledException ex) when (ExceptionUtilities.IsCurrentOperationBeingCancelled(ex, linkedTokenSource.Token))
+            {
+                return null;
+            }
+        }
+        else
+        {
+            // Normal request to either compute frozen partial tags, or compute normal tags in a tagger that does
+            // *not* support frozen partial tagging.
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/NavigationBar/NavigationBarController_ModelComputation.cs:61]:
```suggestion
            // Normal request to either compute nav-bar items using frozen partial semantics.
```

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski ptal.

File: src/EditorFeatures/Core/NavigationBar/NavigationBarController.cs [hunk 4]
```diff
@@ -64,7 +65,15 @@ internal partial class NavigationBarController : IDisposable
...
-    private readonly AsyncBatchingWorkQueue<VoidResult, NavigationBarModel?> _computeModelQueue;
+    private readonly AsyncBatchingWorkQueue<NavigationBarQueueItem, NavigationBarModel?> _computeModelQueue;
+
+    /// <summary>
+    /// This cancellation series controls the non-frozen nav-bar computation pass.  We want this to be separately
+    /// cancellable so that if new events come in that we cancel the expensive non-frozen nav-bar pass (which might be
+    /// computing skeletons, SG docs, etc.), do the next cheap frozen-nav-bar-pass, and then push the
+    /// expensive-nonfrozen-nav-bar-pass to the end again.
+    /// </summary>
+    private readonly CancellationSeries _nonFrozenComputationCancellationSeries;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/NavigationBar/NavigationBarController.cs:76]: same pattern as tagging (including nameing).

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski @sharwell  @ToddGrun ptal.  This impacts a large partner.

File: src/EditorFeatures/Core/NavigationBar/NavigationBarController.cs [hunk 5]
```diff
@@ -92,11 +101,12 @@ internal partial class NavigationBarController : IDisposable
         _visibilityTracker = visibilityTracker;
         _uiThreadOperationExecutor = uiThreadOperationExecutor;
         _asyncListener = asyncListener;
+        _nonFrozenComputationCancellationSeries = new(threadingContext.DisposalToken);
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/NavigationBar/NavigationBarController.cs:104]: We don't have a token for when a file is closed?

File: src/EditorFeatures/Core/NavigationBar/NavigationBarController.cs [hunk 6]
```diff
@@ -196,7 +206,13 @@ private void StartModelUpdateAndSelectedItemUpdateTasks()
         if (_disconnected)
             return;
 
-        _computeModelQueue.AddWork(default(VoidResult));
+        // Cancel any expensive, in-flight, nav--bar work as there's now a request to perform lightweight tagging. Note:
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/NavigationBar/NavigationBarController.cs:209]:
```suggestion
        // Cancel any expensive, in-flight, nav-bar work as there's now a request to perform lightweight tagging. Note:
```

File: src/EditorFeatures/Core/NavigationBar/NavigationBarController_ModelComputation.cs [hunk 7]
```diff
@@ -43,6 +85,12 @@ internal partial class NavigationBarController
 
         var model = await ComputeModelAsync().ConfigureAwait(false);
 
+        // If we were computing with frozen partial semantics here, enqueue work to compute *without* frozen
+        // partial snapshots so we move to accurate results shortly. Create and pass along a new cancellation
+        // token for this expensive work so that it can be canceled by future lightweight work.
+        if (frozenPartialSemantics)
+            _computeModelQueue.AddWork(new NavigationBarQueueItem(FrozenPartialSemantics: false, _nonFrozenComputationCancellationSeries.CreateNext(default)));
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/NavigationBar/NavigationBarController_ModelComputation.cs:92]: It seems like this should just live in the caller -- you already have the if/else if, and then that way there is only one function doing queue management, and one function doing actual data computation.

**Review by jasonmalinowski — APPROVED** [review]
It feels like there should be an easier way to do this. If a feature was added to the queue to just say "and run this once there are no batches left" it might be easier? I think? Either way since this pattern is established, no reason to try to delay this here.

File: src/EditorFeatures/Core/NavigationBar/NavigationBarController_ModelComputation.cs (hunk 7 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/NavigationBar/NavigationBarController_ModelComputation.cs:92]: sgtm!

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/NavigationBar/NavigationBarController_ModelComputation.cs]: Thanks.  that was cleaner!

**Comment by CyrusNajmabadi** [comment]
> It feels like there should be an easier way to do this. If a feature was added to the queue to just say "and run this once there are no batches left" it might be easier?

I'm waiting for the 3rd feature to need this.  At that point, i'll wrap up in a helper (or make part of the queue itself :)).

File: src/EditorFeatures/Core/NavigationBar/NavigationBarController.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/NavigationBar/NavigationBarController.cs]: you are correct.  and i *keep* making that mistake!

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/NavigationBar/NavigationBarController.cs]: (basically, i often see the IThreadingContext and just say in my 'this has the same lifetime as that').

**Pull request merged**